designs/2022-async-parse/README.md

```diff
@@ -0,0 +1,238 @@
+* `it` must support returning a Promise from passed callbacks. It should await for the Promise to be settled and treat the test as "passing" only if the Promise is fulfilled. If the Promise is not settled within a certain timeframe or is rejected, the test should be treated as "failing". Mocha and Jest already support this. 
+
+Projects that use `RuleTester` with Mocha or Jest wouldn't be required to change anything.
+
+Projects that run tests directly or use a testing framework that doesn't support returning a Promise from callbacks will have to switch to a framework like Mocha or Jest.
+
+## Alternatives
+
+### Alternatives to `async parse()` / `async parseForESLint()`
+
+If the use of compiled parsers with ESLint is the only use case for `async parse` / `async parseForESLint` we are aware of, then the question is - is `async parse` / `async parseForESLint` the only way to enable it? In particular, if the parser only needs a one-time async initialization step (I believe that would be the case with Wasm), then the async initialization can be performed in `eslint.config.js` on loading, since the new flat config system supports [async function configs](https://github.com/eslint/rfcs/tree/main/designs/2019-config-simplification#can-a-config-be-an-async-function).
+
+### Alternatives to async-only `Linter`
+
+Instead of removing support for synchronous linting, `verify()` and `verifyAndFix()` could return results synchronously when the parser is sync, or we could add separate methods `verifySync()` and `verifyAndFixSync()`.
+
+This would add a lot of complexity as we would have to maintain both sync and async code paths throughout the `Linter`. It would become even more complicated when we add other async features, like async rules.
+
+### Alternatives to requiring a testing framework
+
+Instead of throwing an error if custom/global `it` was not found, `RuleTester#run` could run all test directly and return a Promise, but that could make the behavior of this method quite confusing for users.
+
+We could add a separate method `RuleTester#runDirectly` that always runs the tests directly and always returns a Promise, but the question is whether anyone needs it.
+
+## Open Questions
+
+* Do we want to explore parallel parsing-linting as described in https://github.com/eslint/eslint/discussions/14733 in this RFC or separately? 
+* Assuming that we plan to implement both `async parse()` support and Flat Config for v9.0.0, can we split `FlatLinter` out of the `Linter` now? The work on `async parse()` overlaps a lot with the work on flat config. With the current dual-mode `Linter`, it doesn't seem feasible to start implementing support for `async parse()` before the final v9.0.0 flat config changes are merged. If we split out `FlatLinter`, most of the work could be finished much earlier and we could have a working support for async parsers in the "flat" mode (the mode that uses `FlatESLint` class).
```

Comment:
This is an idea worth exploring. I’m a bit concerned about the effects on the ecosystem of changing formerly sync methods to be async. We don’t really know all the places that Linter is used, so such a change could break a lot of places we just aren’t aware of yet. (Maybe @bmish could modify his GitHub search he used for rule formats to see where Linter is being used?)

Tying the new async support to flat config could be a nice transitional phase to smooth over this change. There would be the open question of how long to keep the old Linter class around in this case. 

Note that we ultimately don’t want the word “flat” in any interface names as flat config will be the only config eventually.

Reply:
My guess would be that it makes it prohibitively difficult to maintain an eslint plugin that’s compatible with both 8 and 9 (assuming 9 is what adds this).

Reply:
@ljharb do you know of any eslint plugins that use the `Linter` class? That sounds like a very unusual use case to me.

Reply:
- https://github.com/sarbbottam/eslint-find-rules/blob/6dabfb759697e7d946bcfd485be0b7f584f2c0a0/src/lib/rule-finder.js#L74
 - https://github.com/yannickcr/eslint-plugin-react/blob/9e12d2b5054230e366f18fdaa9ad44d952a4b950/tests/lib/rules/jsx-uses-react.js#L39

are all i found in a quick search.

Reply:
Keep in mind that the ecosystem is more than just plugins. We need to consider other tools that may wrap ESLint in some way, too.

Reply:
> * https://github.com/sarbbottam/eslint-find-rules/blob/6dabfb759697e7d946bcfd485be0b7f584f2c0a0/src/lib/rule-finder.js#L74
> * https://github.com/yannickcr/eslint-plugin-react/blob/9e12d2b5054230e366f18fdaa9ad44d952a4b950/tests/lib/rules/jsx-uses-react.js#L39
> 
> are all i found in a quick search.

`Linter#getRules` and `Linter#defineRule` wouldn't be affected by this change (they will be removed by https://github.com/eslint/eslint/issues/13481 though, but that's off-topic for this RFC).

I should have been more precise, the question is which plugins and tools use `Linter#verify` or `Linter#verifyAndFix`, as those are the only public methods that will be changed by this RFC. The change would be especially problematic for plugins because rules and processors cannot be async.

Reply:
In the first several pages of a GitHub search for [`linter.verify()`](https://github.com/search?q=linter.verify()&type=Code) in code, all of the results are in test files. There are a few instances of [`linter.verifyAndFix()`](https://github.com/search?q=linter.verifyAndFix()&type=Code) outside of tests.